### PR TITLE
feat: implement in-process-flagd provider

### PR DESCRIPTION
**Pull request by bacherfl**
This PR adds the in-process flagd provider, following the [flagd specification](https://github.com/open-feature/flagd/blob/main/docs/other_resources/in-process-providers/specification.md)

### CODE REVIEW

**Review by odubajDT — APPROVED** [review]
Nice work! :)

File: providers/in-process-flagd/provider.go
Comment (toddbaert) [inline providers/in-process-flagd/provider.go:11]: really great that we can use these libs. The implementation is surprisingly small.

File: providers/in-process-flagd/provider.go
Comment (toddbaert) [inline providers/in-process-flagd/provider.go:155]:
```suggestion
func (p *Provider) startSyncSource(dataSync chan sync.DataSync) {
```

nit: seems like this is just one source, so perhaps this is a better name?

File: providers/in-process-flagd/README.md
Comment (toddbaert) [inline providers/in-process-flagd/README.md:20]: I don't think these caching options are relevant to in-process providers.

File: providers/in-process-flagd/provider.go
Comment (toddbaert) [inline providers/in-process-flagd/provider.go:178]: As of right now, we don't expect any messages other than `sync.ALL`. We may want to log some debug or warning messages if we get others - cc @Kavindu-Dodan

File: providers/in-process-flagd/provider.go
Comment (toddbaert) [inline providers/in-process-flagd/provider.go:194]:
```suggestion
			fmt.Errorf("error resyncing source: %w", err),
```

File: providers/in-process-flagd/provider.go
Comment (toddbaert) [inline providers/in-process-flagd/provider.go:380]: This is a good point... I wonder if we should delay the actual release of this provider until spec `v0.7.0` is implemented in Go (there wouldn't be too much work required).

Comment (toddbaert) [reply providers/in-process-flagd/provider.go:380]: cc @Kavindu-Dodan @mowies  - we may want to consider something similar for Java/JS, neither of them have `STALE` yet (though there's a [draft PR](https://github.com/open-feature/js-sdk/pull/541) for JS)

**Comment by Kavindu-Dodan** [comment]
@bacherfl I think you forgot to include the `go.mod` file. 

Given this is a new module, we need to configure release please configuration [^1] and set manifest [^2] with initial release version. 

Regarding module name 🤔 shall we prefix this with `flagd` (flagd-in-memory). This should help the folder ordering in provider directory [^3] and give the idea that this is a flagd module.

[^1]: https://github.com/open-feature/go-sdk-contrib/blob/main/release-please-config.json
[^2]: https://github.com/open-feature/go-sdk-contrib/blob/main/.release-please-manifest.json
[^3]: https://github.com/open-feature/go-sdk-contrib/tree/main/providers

File: providers/in-process-flagd/Makefile
Comment (Kavindu-Dodan) [inline providers/in-process-flagd/Makefile:3]: 🤔 not sure this was added by mistake, but source doesn't exisit

File: providers/in-process-flagd/README.md
Comment (Kavindu-Dodan) [inline providers/in-process-flagd/README.md:9]: Naming. we can go for either `inprocessflagd` or `flagdInProcess` 🤔

Any opinion @thisthat @toddbaert  @mowies @odubajDT  ?

File: providers/in-process-flagd/README.md
Comment (Kavindu-Dodan) [reply providers/in-process-flagd/README.md:20]: yeah, this should nearly match with Java provider configurations - https://github.com/open-feature/java-sdk-contrib/tree/main/providers/flagd#configuration-options

File: providers/in-process-flagd/provider.go
Comment (Kavindu-Dodan) [reply providers/in-process-flagd/provider.go:178]: yeah, +1 to log other states.

File: providers/in-process-flagd/provider.go
Comment (Kavindu-Dodan) [reply providers/in-process-flagd/provider.go:380]: Given that this is part of the spec [^1], state `STALE` should be added to SDK. 

[^1]: https://openfeature.dev/specification/types#provider-events

Comment (Kavindu-Dodan) [reply providers/in-process-flagd/provider.go:380]: ohh my bad, provider event stale is not yet reflected in provider state. This PR need to be merged for this [^1]

[^1]: https://github.com/open-feature/spec/pull/200

Comment (toddbaert) [reply providers/in-process-flagd/provider.go:380]: Stale state actually is in the spec `main`, just not release or implemented in very many SDKs: https://github.com/open-feature/spec/blob/main/specification/types.md#provider-status

This will come with spec `0.7.0` release. I'll try to do that tomorrow: https://github.com/open-feature/spec/issues/202

File: providers/in-process-flagd/README.md
Comment (thisthat) [reply providers/in-process-flagd/README.md:9]: I favor the first one: `inprocessflagd`, since it reads better. 
Some reading for guidance on pkg naming: https://go.dev/blog/package-names

File: providers/in-process-flagd/provider.go
Comment (thisthat) [reply providers/in-process-flagd/provider.go:216]: or we could stack the recovery and the unlock in two defer calls at the top level. 
```go
p.mu.Lock()
defer func() {
  if recoveryFunc != nil {
	<-time.After(p.connectionInfo.backoffDuration)
	recoveryFunc()
  }
}()
defer p.mu.Unlock()
```

Mind the order of defer: https://go.dev/blog/defer-panic-and-recover

Comment (thisthat) [reply providers/in-process-flagd/provider.go:216]: I just realized that the defer option wouldn't work because of the retry logic 😭

File: providers/in-process-flagd/provider.go
Comment (Kavindu-Dodan) [inline providers/in-process-flagd/provider.go:29]: Regarding max retries 🤔 

For flagd remote evaluations, we can define max retries for sync and go offline when limit is reached. This is because we always try to evaluate flags against remote deployment of flagd. So even if sync is shutoff, evaluation can reach flagd when flagd is back online. 

But for in-process evaluations, our flag definitions comes from remote sync service. If this connection breaks, we have to continuously try to connect to the remote flag source. Otherwise, our store will be stale and without a process restart (application restart) we cannot refresh flags. I think this should be added to flagd specification 

cc - @toddbaert & @thisthat

Comment (thisthat) [reply providers/in-process-flagd/provider.go:29]: That's a valid point @Kavindu-Dodan 👍

File: providers/in-process-flagd/internal/logger/logger.go
Comment (bacherfl) [reply providers/in-process-flagd/internal/logger/logger.go:26]: it turns out that right now the internal package is not needed anyway - this was a copy-paste relict from the original flagd provider which this one was based off

File: providers/in-process-flagd/provider.go
Comment (bacherfl) [reply providers/in-process-flagd/provider.go]: agree @Kavindu-Dodan , that's a good point - I will adapt the implementation to keep retrying in case of an error

Comment (bacherfl) [reply providers/in-process-flagd/provider.go]: if we keep retrying - do you think it still makes sense to eventually go from `STALE` into `ERROR` after a certain amount of retries, or should we get rid of the `MAX_SYNC_RETRIES` env var altogether, and stay in `STALE` for cases where we previously were able to retrieve the flag configs but ran into a interrupted connection afterwards?

Comment (Kavindu-Dodan) [reply providers/in-process-flagd/provider.go]: Good point. For Java provider, I went ahead with status `STALE` and set a max backoff [^1]. If we agree, we can set status to error once we hit the max backoff 🤔 

[^1]: https://github.com/open-feature/java-sdk-contrib/blob/main/providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/resolver/process/storage/connector/grpc/GrpcStreamConnector.java#L29-L30

Comment (bacherfl) [reply providers/in-process-flagd/provider.go]: Thank you @Kavindu-Dodan - i think the way you implemented that in the Java library makes the most sense here, so I will apply this logic in this provider as well :)

Comment (toddbaert) [reply providers/in-process-flagd/provider.go]: I think if we use the max backoff, it makes sense to make that the "boundary" between STALE and ERROR - but TBH I'm not sure in a server context that anyone would "be okay" with either situation for any substantial amount of time.

I guess a STALE state is tolerable for brief periods and could be characterized as eventual consistency, whereas after max backoff you're at a threshold you've determined to be an error and you might want error-related telemetry to be recorded.

**Review by toddbaert — APPROVED** [review]
Approved pending comment [here](https://github.com/open-feature/go-sdk-contrib/pull/327#discussion_r1321695112). I'll add the test suite once this is merged and then we can consider what we'd want before a final release (we may want spec `0.7.0` implemented first).

File: providers/in-process-flagd/provider.go
Comment (bacherfl) [reply providers/in-process-flagd/provider.go]: Thanks @toddbaert , that makes sense - that's essentially also how the provider is implemented right now. So now this provider will keep retrying with the max backoff being the upper limit, and after exceeding max_retries an error event is sent

File: providers/in-process-flagd/Makefile
Comment (bacherfl) [reply providers/in-process-flagd/Makefile]: yes, good catch, that was indeed added by accident - removed it

File: providers/in-process-flagd/provider.go
Comment (Kavindu-Dodan) [reply providers/in-process-flagd/provider.go]: Yes, it makes sense to emit an error if we reach the max back-off limit. I will add this to Java provider

**Review by Kavindu-Dodan — APPROVED** [review]

**Pull request merged**
